craft_providers/multipass/multipass.py

```diff
@@ -0,0 +1,382 @@
+            raise MultipassError(
+                brief=f"Failed to start VM {instance_name!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def stop(self, *, instance_name: str, delay_mins: int = 0) -> None:
+        """Stop VM instance.
+
+        :param instance_name: the name of the instance_name to stop.
+        :param delay_mins: Delay shutdown for specified number of minutes.
+
+        :raises MultipassError: on error.
+        """
+        command = ["stop"]
+
+        if delay_mins != 0:
+            command.extend(["--time", str(delay_mins)])
+
+        command.append(instance_name)
+
+        try:
+            self._run(command, capture_output=True, check=True)
+        except subprocess.CalledProcessError as error:
+            raise MultipassError(
+                brief=f"Failed to stop VM {instance_name!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def transfer(self, *, source: str, destination: str) -> None:
+        """Transfer to destination path with source IO.
+
+        :param source: The source path, prefixed with <name:> for a path inside
+            the instance.
+        :param destination: The destination path, prefixed with <name:> for a
+            path inside the instance.
+
+        :raises MultipassError: On error.
+        """
+        command = ["transfer", source, destination]
+
+        try:
+            self._run(command, capture_output=True, check=True)
+        except subprocess.CalledProcessError as error:
+            raise MultipassError(
+                brief=f"Failed to transfer {source!r} to {destination!r}.",
+                details=errors.details_from_called_process_error(error),
+            ) from error
+
+    def transfer_destination_io(
+        self, *, source: str, destination: io.BufferedIOBase, chunk_size: int = 4096
+    ) -> None:
+        """Transfer from source file to destination IO.
+
+        Note that this can't use std{in,out}=open(...) due to LP #1849753.
+
+        :param source: The source path, prefixed with <name:> for a path inside
+            the instance.
+        :param destination: An IO stream to write to.
+
+        :raises MultipassError: On error.
+        """
+        command = [str(self.multipass_path), "transfer", source, "-"]
+        proc = subprocess.Popen(
+            command,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+        )
+
+        # Should never happen, but pyright makes noise.
+        assert proc.stdout is not None
+
+        while True:
+            data = proc.stdout.read(chunk_size)
+            if not data:
+                break
+
+            destination.write(data)
+
+        try:
+            _, stderr = proc.communicate(timeout=30)
+        except subprocess.TimeoutExpired:
+            proc.kill()
+            _, stderr = proc.communicate()
```

Comment:
How this `communicate` work after the process was killed?

Reply:
Best practice according to docs.
https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

Reply:
As we discussed in mattermost, this will be replaced/simplified in the next branch.